modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java

```diff
@@ -858,10 +858,14 @@ private void enlist(int tableId, List<NodeWithTerm> assignments) {
                     tx.assignCommitPartition(new TablePartitionId(tableId, ThreadLocalRandom.current().nextInt(partsCnt)));
 
                     for (int p = 0; p < partsCnt; p++) {
-                        NodeWithTerm enlistmentToken = assignments.get(p);
+                        TablePartitionId tablePartId = new TablePartitionId(tableId, p);
 
-                        tx.enlist(new TablePartitionId(tableId, p),
-                                new IgniteBiTuple<>(topSrvc.getByConsistentId(enlistmentToken.name()), enlistmentToken.term()));
+                        if (tx.enlistedNodeAndTerm(tablePartId) == null) {
```

Comment:
It's not that was expected.
First of all proposed solution isn't thread safe. However it doesn't matter because enlist() itself won't re-enlist the partition. The main point here was in a fact that we shouldn't re-await primary replicas if they were already enlisted.

Reply:
this check is appended just to avoid second call on already enlisted partitions. And need to be removed (as mentioned) in https://issues.apache.org/jira/browse/IGNITE-20925